language/proto/gen/update_proto_csv.go

```diff
@@ -66,25 +72,225 @@ func main() {
 	flag.Parse()
 
 	if *protoCsvPath == "" {
-		log.Fatal("-proto must be set")
-	}
-	if *googleapisRootPath == "" {
-		log.Fatal("-go_googleapis must be set")
+		log.Fatal("-proto_csv must be set")
 	}
 
 	protoContent := &bytes.Buffer{}
 	protoContent.WriteString(prefix)
 
-	err := filepath.Walk(*googleapisRootPath, func(path string, info os.FileInfo, err error) error {
+	if *goGoogleapisRootPath != "" {
+		if err := generateFromPath(*goGoogleapisRootPath, protoContent); err != nil {
+			log.Fatal(err)
+		}
+	} else if *comGoogleGoogleapisRootPath != "" {
+		if err := generateFromQuery(*bazelPath, *comGoogleGoogleapisRootPath, protoContent); err != nil {
+			log.Fatal(err)
+		}
+	} else {
+		log.Fatal("either -com_google_googleapis or -go_googleapis must be set")
+	}
+
+	if err := ioutil.WriteFile(*protoCsvPath, protoContent.Bytes(), 0666); err != nil {
+		log.Fatal(err)
+	}
+}
+
+//
+// Process -com_google_googleapis case
+//
+const repoPrefix = "@com_google_googleapis"
+
+var bazelQuery = []string{
+	`query`,
+	`kind("proto_library rule", //google/...)`,
+	`--output`,
+	`xml`,
+	`--experimental_enable_repo_mapping`,
+}
+
+type Query struct {
+	XMLName xml.Name `xml:"query"`
+	Version string   `xml:"version,attr"`
+	Rules   []Rule   `xml:"rule"`
+}
+
+type Rule struct {
+	XMLName xml.Name `xml:"rule"`
+	Class   string   `xml:"class,attr"`
+	Name    string   `xml:"name,attr"`
+	Strings []String `xml:"string"`
+	Labels  []Label  `xml:"label"`
+	Lists   []List   `xml:"list"`
+}
+
+type List struct {
+	XMLName xml.Name `xml:"list"`
+	Name    string   `xml:"name,attr"`
+	Labels  []Label  `xml:"label"`
+}
+
+type Label struct {
+	XMLName xml.Name `xml:"label"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+type String struct {
+	XMLName xml.Name `xml:"string"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+func generateFromQuery(bazelPath, workspacePath string, protoContent io.Writer) error {
+	xmlData, err := runBazelQuery(bazelPath, workspacePath)
+	if err != nil {
+		return err
+	}
+
+	query, err := readQueryXml(xmlData)
+	if err != nil {
+		return err
+	}
+
+	relPathList, err := processQuery(query)
+	for _, v := range relPathList {
+		if _, err = fmt.Fprintf(protoContent, "%s\n", v); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func runBazelQuery(bazelPath, workspacePath string) ([]byte, error) {
+	cmd := exec.Command(bazelPath, bazelQuery...)
+	cmd.Dir = workspacePath
+	return cmd.Output()
+}
+
+func readQueryXml(xmlData []byte) (Query, error) {
+	if bytes.HasPrefix(xmlData, []byte(`<?xml version="1.1"`)) {
+		copy(xmlData, []byte(`<?xml version="1.0"`))
+	}
+	query := Query{}
+	err := xml.Unmarshal(xmlData, &query)
+	return query, err
+}
+
+func processQuery(query Query) ([]string, error) {
+	protoLabelMap := make(map[string]string)
```

Comment:
I'm a little uncomfortable with storing partially rendered information in here. Could you define a new type to use as the value? I think that would make this more readable. Something like:

```
type goProtoLibraryInfo struct {
  label, importpath string
}
```

Also, please add a comment explaining this maps `proto_library` labels to information about `go_proto_library` rules. I don't think that's clear from the name or the type right now.

Reply:
Added